src/Interpreters/ProcessList.cpp

```diff
@@ -313,6 +311,9 @@ ProcessListEntry::~ProcessListEntry()
         }
     }
 
+    if (auto query_user = parent.queries_to_user.find(query_id); query_user != parent.queries_to_user.end())
+        parent.queries_to_user.erase(query_user);
+
```

Comment:
I may be wrong, but at first glance we should erase from queries_to_user after cancelled_cv.wait(), otherwise we are going to have a race condition here.